Racing Game

TonyKim9401 opened the pull request:
Much better than calculator

TonyKim9401 committed:
feat

  before clean code

TonyKim9401 committed:
feat

  after clean code

TonyKim9401 committed:
chore

  add comment and refactor

TonyKim9401 committed:
chore

  package distribute

TonyKim9401 committed:
chore

  in the middle of re-coding

liquidjoo commented on src/main/java/org/step2/Main.java:
```diff
@@ -4,17 +4,25 @@
 
+        // 레이싱 게임 진행
+        RacingCars cars = new RacingCars();
+        cars.initializeRacingCar(carInput,new RandomCondition());
```
입력 값이 초기화 되었다면 생성자를 활용해 볼 수 있겠네
빈 생성자를 사용하는 것은 어떤 단점이 있을까?

liquidjoo commented on src/main/java/org/step2/car/CarLog.java:
```diff
@@ -0,0 +1,24 @@
+package org.step2.car;
+
+public class CarLog implements Log {
```
Log 라는 인터페이스가 꼭 있어야할까?

liquidjoo commented on src/main/java/org/step2/car/CarLog.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public String getLog() {
+        return this.movement;
+    }
```
상태 값을 반환하는 Get 을 사용하는 것도 괜찮아 보임

liquidjoo commented on src/main/java/org/step2/car/RacingCar.java:
```diff
@@ -1,24 +1,22 @@
-    private int moveCount;
     private final Condition condition;
-    private final Output output;
+    private final List<Log> carLogs = new ArrayList<>();
```
기록 시 몇 번째 자동차인지 어떻게 알 수 있을까?

liquidjoo commented on src/main/java/org/step2/car/RacingCars.java:
```diff
@@ -20,15 +19,12 @@ public class RacingCars {
-            this.cars.add(new RacingCar(this.condition, this.output));
+    public void initializeRacingCar(Input input, Condition condition) {
+        for (int i = 0; i < input.getInputValue(); i++) {
+            this.cars.add(new RacingCar(condition));
```
Input 의존성은 미리 입력 값을 받아서 객체로 넘긴다면 해결할 수 있을 것 같네!

liquidjoo commented on src/main/java/org/step2/car/Log.java:
```diff
@@ -0,0 +1,6 @@
+package org.step2.car;
+
+public interface Log {
+
+    String getLog();
+}
```
로그를 인터페이스로 한 이유는?

liquidjoo commented on src/main/java/org/step2/car/RacingCars.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public List<Car> getCars() {
+        return this.cars;
```
방어적 복사에 대해서 한번 알아보면 좋을듯

liquidjoo commented on src/main/java/org/step2/car/RacingCar.java:
```diff
@@ -0,0 +1,33 @@
+     * condition 에 따라 이동/미이동이 결정 됨
+     */
+    @Override
+    public void moveForward() {
+        if (this.condition.getConditionValue()) {
+            this.moveCount += 1;
+        }
+        this.carLogs.add(new CarLog(this.moveCount));
+    }
+}
```
해당 코드를 테스트한다면 어떻게 해볼 수 있을까?

liquidjoo commented on src/main/java/org/step2/condition/RandomCondition.java:
```diff
@@ -0,0 +1,25 @@
+    private boolean condition;
+
+    // Random 은 static 으로 만들어 두고 하나만 사용해도 될 것 같은데..
+    private final Random random;
```
랜덤 객체의 성질을 활용해보면 좋을듯
>static 을 사용하면 좋은점

liquidjoo commented on src/main/java/org/step2/execution/Execution.java:
```diff
@@ -0,0 +1,8 @@
+
+import org.step2.input.Input;
+
+public interface Execution {
+
+    void getExecutionResult(Input input);
+}
```
인터페이스로 만든 이유는?

liquidjoo commented on src/main/java/org/step2/execution/RacingExecution.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class RacingExecution implements Execution {
+
+    private final RacingCars cars;
```
3대의 레이싱카가 있다면 객체를 어떻게 표현할 수 있을까?

liquidjoo reviewed (changes requested):
한번에 고치기는 어려울 것 같아서 조금씩 피드백을 반영해보면 좋을듯
다음 요청 시엔 테스트 코드를 꼭 작성해서 오면 좋을듯!

TonyKim9401 committed:
chore

  need to implement the last part

TonyKim9401 committed:
chore

  improved a bit

TonyKim9401 committed:
feat

  add test code

TonyKim9401 committed:
사실 사용을 하지 않아 삭제

TonyKim9401 committed:
테스트 코드에서 검증 값 확인 용도로 추가

TonyKim9401 committed:
자바 시스템 내부의 출력값을 미리 셋팅하고 RandomCondition 을 목킹하여 원하는 출력 결과가 나오는지 확인

TonyKim9401 committed:
레이싱 게임 실행에 대한 검증인데, 테스트 메소드의 반환 값은 void 이고 내부 로직은 RacingCars 객체와 CarLog 에 연관되어 있어 어떻게 테스트를 하면 좋을지 고민됨. 역할 부여가 제대로 되지 않은 생각도 들고, 실행시 제대로 실행 된건지 안된건지를 판단하는 기준이 불명확한것 같음

TonyKim9401 committed:
CarLog 기록시 moveCount 가 음수인 경우가 존재하지 않지만, 요구사항 중 후진하는 케이스가 없으므로 검증 메소드 추가

CarLog 테스트에 검증 메소드에 대한 경계값 테스트 추가

CarLog 테스트를 위해 public 메소드 getMoveExpression 추가

TonyKim9401 committed:
CarLog 의 LogSequence 저장을 위한 logSequence 필드 추가

TonyKim9401 committed:
CarLog 테스트 코드 오타 수정 및 추가된 logSequence 정보 추가

moveCount 및 logSequence 는 0 미만이 될 경우가 없는데 굳이 검증 메소드를 추가하고 테스트 했어야 했을까? 없어도 괜찮지 않을까?

TonyKim9401 committed:
CarOutput 내부에서 사용하는 RacingCars 의 메소드 이름을 구체적으로 변경

TonyKim9401 committed:
CarOutput 내부에서 사용하는 RacingCars 의 메소드 이름을 구체적으로 변경

TonyKim9401 committed:
RacingCar 클래스에 순서 저장용 carSequence 추가 및 차에 따라 실행 횟수를 저장하는 racingExecuteSequence 추가

RacingCar 는 외부에서 생성 되므로 carSequence 를 생성자를 통해 초기화 하도록 수정

moveForward 메소드 실행시 필요한 비즈로직을 메소드 단위로 분리

TonyKim9401 committed:
RacingCarTest 는 수정된 RacingCar 클래스의 비지니스 로직을 검증함

TonyKim9401 committed:
RacingCarsTest 에서 RacingCars 클래스의 생성자, 각 RacingCar 요소들의 로그를 확인하는 getCarLogsByTryCount 메소드를 테스트 한다.

TonyKim9401 committed:
테스트 코드의 Display 를 현재형으로 변경

TonyKim9401 committed:
CarLog 의 일급 객체인 CarLogs 추가 및 테스트 코드 작성

TonyKim9401 committed:
RacingCar 에서 기존 사용하던 CarLog 리스트를 일급 객체로 변경해서 사용하여 응집성과 책임을 좀 더 명확히 나눔

liquidjoo commented on src/main/java/org/step2/car/CarLog.java:
```diff
@@ -1,24 +1,40 @@
 
-    public CarLog(Integer moveCount) {
+    public CarLog(Integer logSequence, Integer moveCount) {
+        this.logSequence = logSequence;
+        validateMoveCount(moveCount);
```
```suggestion
        validateMoveCount(moveCount);
        this.logSequence = logSequence;
```
이런게 더 낫긴하지

TonyKim9401 replied:
검증 먼저 진행 해야 하는 부분을 진행 하고 통과 되면 다른 작업 처리하는 거군요..

liquidjoo commented on src/main/java/org/step2/car/CarLogs.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * CarLogs 의 응집성을 위해 일글 컬렉션으로 처리
+ */
+public class CarLogs {
```
CarLogs 라는 네이밍보다 Cars 를 쓰면 되지 않을까??

TonyKim9401 replied:
CarLog의 일급 컬렉션인데 이름을 Cars로 하는게 나은가요?

liquidjoo commented on src/main/java/org/step2/car/RacingCar.java:
```diff
@@ -1,33 +1,71 @@
+    private Integer carSequence;
+    private Integer position;
+    private Integer racingExecuteSequence;
+    private CarLogs carLogs;
```
RacingCar 가 CarLogs 를 들고 있는게 맞을까?
DIP 에 대해서도 공부해보면 좋을듯

TonyKim9401 replied:
RacingCar 에 대한 기록이라고 생각해서 클래스 내부에 넣는게 맞다고 생각 했는데 DIP를 생각 해서 리팩토링 한다면 CarLogs 자체를 추상화 해서 RacingCar 에서 주입 받는 식으로 변경 하는 방향으로 가야 하는 걸까요? 상위 객체가 하위 객체에 너무 의존하는 생각은 들긴 합니다만 개인적으로 딜레마가 조금 있는것 같습니다.

liquidjoo commented on src/main/java/org/step2/car/RacingCar.java:
```diff
@@ -1,33 +1,71 @@
-        this.carLogs.add(new CarLog(this.moveCount));
+        checkMoveCondition();
+        recordCarLog();
+        increaseRacingExecuteSequence();
```
단일책임원칙을 생각해본다면?

liquidjoo commented on src/main/java/org/step2/car/RacingCar.java:
```diff
@@ -1,33 +1,71 @@
+        return this.carSequence;
+    }
+
+    /**
+     * 메소드 이름을 똑같이 해두면 사용시 헷갈릴것 같은데 같은 목적을 가졌을 때 어떻게 명명하면 좋을까?
+     */
```
같은 목적일때 메서드가 2개가 필요한가?
추상화를 생각해보거나 메서드 명을 더 명확하게 만들어보거나

liquidjoo commented on src/main/java/org/step2/car/RacingCars.java:
```diff
@@ -1,34 +1,39 @@
-     * initializeRacingCar 메소드를 사용해
-     * output 에서 input 값을 확인하여 car 를 초기화 한다.
-     * -> 의존성이 너무 복잡하게 되는것 같은 생각이 든다.
+     * for 문 사용으로 인덱스 i를 사용해 자동차의 번호를 정한다.
+     * -> stream 사용 안한 이유
+     * 미리 입력 값을 받아서 객체로 넘긴다는게 Integer 값을 직접 받는다는게 맞을지?
```
자동차의 번호가 중요한가? 자동차의 이름을 받는다면 어떻게 할 것인가?

TonyKim9401 replied:
1. 자동차의 번호는 상황에 따라 필요할수도 필요하지 않을 수도 있을것 같습니다.
2. 자동차의 이름, 닉네임 등등의 값을 받아야 한다면 생성자를 통해 Input 인터페이스를 입력 받는 식으로 처리할 것 같습니다. 그러려면 RacingCars 생성자를 통해서요 Input 인터페이스를 받아야 할 것 같구요!

liquidjoo commented on src/main/java/org/step2/output/CarOutput.java:
```diff
@@ -1,16 +1,43 @@
+    private final RacingCars racingCars;
+    private final Integer tryCount;
 
+    /**
+     * CarOutput 을 사용 하려면 필수 2가지 파라미터
+     * RacingCars 와 tryCount 가 있어야 한다고 생각하여
+     * 생성자에서 받도록 수정
+     * -> tryCount 는 결국 CarLogs 의 사이즈와 같다고 생각되는데 외부 주입을 받아야 할까?
+     * tryCount 를 명시적으로 하기 위해 외부에서 받는 방향으로 수정해 보았다.
+     */
```
tryCount 를 사용하면 안돼
이미 완성된 객체를 통해서 output을 만들어야지

liquidjoo commented on src/test/java/org/step2/car/RacingCarTest.java:
```diff
@@ -0,0 +1,156 @@
+        RacingCar racingCar = new RacingCar(carSequence, condition);
+
+        // then
+        assertThat(racingCar.checkCarSequence()).isEqualTo(carSequence);
+        assertThat(racingCar.getPosition()).isEqualTo(0);
+        assertThat(racingCar.getCarLogs()).hasSize(0);
```
assertAll 과 어떤차이가 있을까?
assertAll 도 사용해보면 좋음

TonyKim9401 replied:
assertThat 을 사용하면 여러개 실패해도 가장 빨리 실패한 한 가지 케이스만 확인 가능한데, assertAll 을 사용하면 해당되는 실패 케이스를 모두 한번에 확인 가능하군요 :)

liquidjoo reviewed (changes requested):
토니야 고생했다
피드백 남겨놓았으니 한번 봐!

TonyKim9401 committed:
validateMoveCount 를 통해 검증 우선 진행

TonyKim9401 committed:
중복 메소드 이름 변경

TonyKim9401 committed:
CarOutput 출력시 결과 값에만 의존 하도록 변경

TonyKim9401 committed:
assertAll 을 사용하여 테스트에 대한 효율을 높임
